src/kemal/base.cr

```diff
@@ -0,0 +1,119 @@
+require "./helpers/*"
+require "./base/*"
+
+# Kemal Base
+# The DSL currently consists of
+# - get post put patch delete options
+# - WebSocket(ws)
+# - before_*
+# - error
+class Kemal::Base
+  include FileHelpers
+  include Templates
+  include Macros
+  include Base::DSL
+  include Base::Builder
+
+  # :nodoc:
+  getter route_handler = Kemal::RouteHandler.new
+  # :nodoc:
+  getter filter_handler = Kemal::FilterHandler.new
+
+  getter handlers = [] of HTTP::Handler
+  getter websocket_handlers = [] of HTTP::WebSocketHandler
+  getter error_handlers = {} of Int32 => HTTP::Server::Context, Exception -> String
+
+  getter config : Config
+
+  property! logger : Kemal::BaseLogHandler
+  property! server : HTTP::Server
+  property? running = false
+
+  def initialize(@config = Config.base)
+    initialize_defaults
+  end
+
+  def run(port = nil)
+    run(port) { }
+  end
+
+  # Overload of self.run to allow just a block
+  def run(&block)
+    run nil, &block
+  end
+
+  # The command to run a `Kemal` application.
+  # The port can be given to `#run` but is optional.
+  # If not given Kemal will use `Kemal::Config#port`
+  def run(port = nil, &block)
+    setup
+
+    prepare_for_server_start
+
+    start_server port, &block
+  end
+
+  def self.run(port = nil, &block)
+    new.tap do |app|
+      Kemal::CLI.new(app.config)
+
+      app.run(port, &block)
+    end
+  end
+
+  def self.run(port = nil)
+    new.tap do |app|
+      Kemal::CLI.new(app.config)
+
+      app.run(port)
+    end
+  end
+
+  # DEPRECATED: This method should be replaced with `#running?`
+  def running
+    running?
+  end
+
+  private def prepare_for_server_start
+    unless @config.env == "test"
```

Comment:
This casues a crash when an app is initialized with a custom `Kemal::Config` instance